Add automatic versioning

nilsvu opened the pull request:
## Proposed changes

Adds automatic versioning to our CI, so a new release is published every month. A release can also be published manually by [dispatching the workflow from GitHub](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/). The releases look like this: https://github.com/nilsleiffischer/spectre/releases

Note that the script `CompileReleaseNotes.py` is actually also quite useful just for checking what happened to the repository since a particular commit. Try running this (with Py3):
```sh
pip install GitPython
python tools/CompileReleaseNotes.py --from HEAD~10
```

Here's what we'll need to do before merging this PR:

- [x] Choose a date on which we want to release the initial version. Put it in `Metadata.yaml`.
  **Edit:** I selected Dec 7th, 2020 for now.
- [x] Create a draft record on Zenodo so we have a pre-reserved DOI (don't publish yet). Put the DOI into `Metadata.yaml`.
  **Edit**: Here it is: https://doi.org/10.5281/zenodo.4290405. It's not public yet, but the link should start working once we publish it.
- [x] Merge a PR with the `Metadata.yaml` into the repository some time before the initial release date. This PR should be announced so everyone in the author list gets a chance to see it.
  **Edit:** Here it is: #2606 
- [x] Tag the merge commit from the `Metadata.yaml` PR with the initial version name.
- [x] Push the initial version tag to the `release` branch.
- [x] Create a GitHub release from the initial version tag.
- [x] Upload the release zipball to Zenodo. Also update the Zenodo record so it corresponds to the `Metadata.yaml` in the repo, we can use one of the scripts in this PR for that.
- [x] Publish the Zenodo record. The automatic versioning is now ready to take over! We can merge this PR.

At any time:

- [x] Create the labels `bugfix` (self-explanatory, these PRs get moved to the bottom of the release notes) and `major new feature` (to highlight just a few major PRs at the top of the release notes) and retro-actively apply these labels to appropriate PRs merged in November. Then use the labels somewhat consistently from now on. <s>Also consistently apply the `breaking change` label to PRs merged in November, though I'm not sure that label makes a lot of sense, since, as @nilsdeppe has regularly pointed out on these occasions, most changes are breaking for someone and it's even less clear where to draw the line between "breaking" and "non-breaking" than for "bugfix" and "major new feature".</s>

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add automatic versioning

Adds the ability to create a release automatically by running a GitHub
workflow. The workflow can be triggered manually or on a schedule, though
the scheduled trigger is not yet enabled so we have some time to test
the workflow.

wthrowe commented on CMakeLists.txt:
```diff
@@ -3,7 +3,14 @@
 
-# Fortran below is needed for SPHEREPACK.
+# Set up the project. Notes:
+# - Scripts may automatically regex-replace the version string in this command.
+#   For example, here's how you could set the version to 'X.Y.Z':
+#
+#   sed -i 's/^\(project(.*VERSION \)\([^ )]\)*/\1X.Y.Z/' CMakeLists.txt
+#
+#   Please don't break this command when editing this file.
```
I don't understand what this comment is telling me to do.  When should I be setting the version number?

nilsdeppe replied:
I also don't understand what "break" means

nilsvu replied:
clarified in a fixup

wthrowe commented on tools/CompileReleaseNotes.py:
```diff
@@ -0,0 +1,192 @@
+
+if __name__ == "__main__":
+    # The release notes always refer to the repository that contains this file
+    repo = git.Repo(os.path.dirname(os.path.dirname(__file__)))
```
Does the python interface provide something equivalent to `git rev-parse --show-toplevel`?

nilsvu replied:
Yes! Fixed.

wthrowe commented on .github/workflows/Tests.yaml:
```diff
@@ -20,6 +21,18 @@ on:
+  workflow_dispatch:
+    inputs:
+      release_version:
+        description: "Set to create a release on success (no 'v' prefix)"
```
I'm not sure that I'm associating bits of this file correctly, but below I find
```
      # We use the current date as tag name, unless a tag name was specified
      # as input to the `workflow_dispatch` event
￼```
so if this is unset does it create a release with a default name or not create a release?

(I don't know why GitHub is refusing to format this comment.)

nilsvu replied:
<s>It will create the release with the default name.</s> Sorry I confused something. No, it will not create a release if dispatched manually and this is unset. Reason being that we may want to dispatch the workflow without creating a release. Also I want to be safe and only create a release if explicitly requested.

wthrowe commented on CMakeLists.txt:
```diff
@@ -3,7 +3,14 @@
+# - Scripts may automatically regex-replace the version string in this command.
+#   For example, here's how you could set the version to 'X.Y.Z':
+#
+#   sed -i 's/^\(project(.*VERSION \)\([^ )]\)*/\1X.Y.Z/' CMakeLists.txt
```
It looks like this PR moves us away from X.Y.Z versioning, so that's probably a bad example.

wthrowe reviewed (changes requested):
I don't think I've seen anyone advocate for "breaking change" recently.

Is there a reason to have the `release` branch?  I've generally seen projects just use tags without an additional structure.

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +446,86 @@ jobs:
+          echo "RELEASE_VERSION=$RELEASE_VERSION" >> $GITHUB_ENV
+      - name: Validate release version
+        run: |
+          VERSION_PATTERN="^[-.0-9a-zA-Z]+$"
```
I think we can reasonably require that the year and month in the version match our scheme. The date could see some variation, but even there it should be more than a couple of days. Could you add a check to verify the year and month are correct? Also, why would we get letters in the version? I think the required pattern should be: `YYYY.MM.DD(.TWEAK_NUMBER)`. This works with CMake

nilsvu replied:
- Yes to narrowing down the pattern in the regex
- Are you suggesting to require that the year and month in the manually-specified tag match today's year and month? Can't we trust ourselves to use the power to choose the version name responsibly? I want to avoid situations where we want to fix something but have to jump through hoops because we were too restrictive here. For example, if today is Jan 1st and we forgot to release some version before the end of the year, I think we should just be able to backdate it (the branch `HEAD` is the same, after all). I could even add the option to specify the revision to tag, instead of the default `HEAD`.
- Letters are typically used for pre-release versions like `1.2.3-alpha`. Again, I think we should trust ourselves and not be too restrictive here. It's hard to anticipate what we want to do in the future.
- I used dashes instead of dots because I was worried that `YYYY.MM.DD` could be mis-interpreted as semantic versioning, implying some sort of compatibility guarantee. In fact cmake reads the components as "major", "minor" and "patch". Do you think that's not an issue?

nilsdeppe replied:
- :+1:
- Yep, that's what I'm suggesting. I'm not sure we should be backdating tags like that. This will break horribly if we had any PR merged between the backdate and the moment the tag is applied, I think. Or at least I'm not at all convinced this approach is safe. That might just be an issue with allowing a nonlinear history, though. @wthrowe do you have any experience with this? Should we switch to a linear history to make `checkout tag` not do something weird where we get "newer" commits in the tag? I have no experience with this.
- Since we are not using semantic versioning your example doesn't apply. No, I don't think we should trust ourselves. Everything we can check we should check.
- We should use dots, that's super standard. I don't think it's really an issue what CMake calls them, and we can't really change what CMake calls things. In a way it's actually not "too incorrect" what CMake calls them, but luckily we won't need to deal with that ever really anyway aside from filling things into some of the `cpp` files and the documentation, where all we really care about is what values they hold.

nilsvu replied:
Ok to dots. How about we don't zero-pad the month and day then, and even leave out the day on the standard scheduled monthly releases: `v2020.11`? I don't care much if we allow letters in the "tweak" component, I lean towards allowing things like `v2020.11.1.1-alpha` but am fine either way.

On backdating: Say we have `develop` point to commit `A`, which is a merge commit from a PR merge on Dec 31st, 2020. We planned on releasing that commit as `v2020.12.31` for whatever reason. But we forgot to do that on Dec 31st, because, you know, we might not have spent that particular day coding :) So I think on Jan 1st (or later) we should be able to just go on GitHub, click the "Dispatch Workflow" button, put in `v2020.12.31` as the tag name and `A` as the revision and hit "Run" to create that release.

wthrowe replied:
> do you have any experience with this? Should we switch to a linear history to make checkout tag not do something weird where we get "newer" commits in the tag? I have no experience with this.

I have no personal experience, but I think it will be OK without us changing anything.  Going backwards in history from the tag should look normal.  We may get slightly odd-looking things like a tag containing commits that are older than the previous tag, but not contained in it, but that shouldn't actually matter.  If we want to backdate, I think it is still fine as long as the old tag is reachable from the new one, which should be true for any PR merge commit.

> We should use dots

Agreed.  No one will confuse us as being on semver 2000+.

> How about we don't zero-pad the month and day then

Ew.  (And semver doesn't zero-pad, so this would seem to make confusion more likely.)

>  even leave out the day on the standard scheduled monthly releases

This seems reasonable.  We should make sure it's before any manual tags during the month, but I don't see an issue with that.

> On backdating: Say we have develop point to commit A, which is a merge commit from a PR merge on Dec 31st, 2020. We planned on releasing that commit as v2020.12.31 for whatever reason. But we forgot to do that on Dec 31st, because, you know, we might not have spent that particular day coding :) So I think on Jan 1st (or later) we should be able to just go on GitHub, click the "Dispatch Workflow" button, put in v2020.12.31 as the tag name and A as the revision and hit "Run" to create that release.

But why would we want to do that?  Is someone going to rely on the name of a tag we haven't created yet?  There's no reason to write the script to make it easier to do something if there's no reason to do that thing.

nilsvu replied:
Re version format: Do we agree on `YYYY.0M[.0D[.TWEAK]]`? I.e., a standard scheduled release at the beginning of Jan 2021 would be `2021.01`, an extra release on Jan 15th 2021 would be `2021.01.15` and another (unlikely) release on the same day would be `2021.01.15.1`.

nilsdeppe replied:
I agree on the formatting padding with zeros.

I would prefer the standard release including the date so it's clear which day that "standard" is released. Otherwise I worry that it'll be unclear if `2020.10.15` is older or newer than `2020.10` since we could be tagging at the end of the month. While this should all be documented, I don't see the drawback to being explicit here, but maybe I'm missing something.

nilsvu replied:
- By standard versioning rules `2020.10.15` would be newer than `2020.10`
- I expect we will very rarely have extra releases within a month. If feel like the more concise `2020.10` promotes the idea of this being "the October release", which is easy to refer to when talking to people.

nilsvu replied:
@nilsdeppe I'll write up the rules that we end up with, and a description how the release schedule is automated, in a dev guide

nilsdeppe replied:
why can't we just refer to 2020.10.01 as the Oct. release? It's just as easy to say and retains the additional info? I don't think there are "standard versioning rules" really, especially not for by date. Often the extra `.0.0` is included with semantic versioning, so why should we cut out the extra info?

nilsvu replied:
ok sure fine with me

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -20,6 +21,18 @@ on:
   push:
     branches-ignore:
       - gh-pages
+  # Once a month we run the tests and release a new version. The scheduled
```
I don't understand from any of the comments I've read how to specify the tag manually.

Maybe adding a dev guide entry on how to manage releases should be added? I'm fine with waiting to write that until the implementation is ironed out.

nilsvu replied:
added a dev guide

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +446,86 @@ jobs:
+    needs:
+      - check_files_and_formatting
+      - doc_check
+      - unit_tests
```
I'm a bit on the fence in terms of requiring the tests, the only reason being that we do still get occasional random failures. Having said that, it would be good to only tag releases that pass everything. I'm not sure what the right thing to do is. If the code is actually broken we could tag the next day as the release or whenever it's fixed. 

What do others think?

nilsvu replied:
Yes I was wondering about that as well. My suggestion here would be to require the tests pass, and if one fails randomly  we [rerun the workflow](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/). Better safe than sorry :)

nilsdeppe replied:
If I'm understanding correctly you're saying we just manually restart the tagging portion of CI? Or rerun the whole CI?

nilsvu replied:
I think GitHub only allows manually dispatching the whole workflow. So when the scheduled release fails randomly, we click "Dispatch Workflow", it will run the tests again and then the tagging, and probably not fail again (else we should really do something about the random test and not release a failing test).

nilsdeppe commented on tools/CompileReleaseNotes.py:
```diff
@@ -0,0 +1,192 @@
+    if args.github_token:
+        import github
+        import tqdm
+        gh = github.Github(args.github_token)
```
I find it surprising you need a token to retrieve public repo info. Is this just GitHub trying to limit web crawlers or something like that?

nilsvu replied:
Yes. But as it turns out GitHub also allows [60 unauthenticated requests per hour](https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#rate-limiting) based on the IP address, so I'll change the script to allow that.

nilsdeppe replied:
Just to make sure we're on the same page: you're saying you'll make it so you can either specify an access token _or_ hope you haven't hit the 60 per hour limit?

nilsvu replied:
Yes. And add an opt-out CLI option.

nilsdeppe replied:
Okay, great! Thank you!

nilsdeppe reviewed (commented):
Overall this looks really nice. Mostly just some minor points to discuss/get thoughts on.

nilsvu commented:
@wthrowe Re your question about the `release` branch: I thought it can be useful for things like Spack to point to the latest release. I'm using it in the GitHub workflow to abort creating a new release if nothing happened on `develop` since the last release.

wthrowe commented:
OK, a way to get the latest release makes sense.

Do we actually want to abort instead of making two identical releases?  I feel like a manual release near the end of a month shouldn't prevent the main release from happening.  Having the January release called `v2020.12.27a` or whatever seems a bit confusing.

nilsvu commented:
@wthrowe Not sure.. I thought it would be silly to create releases when nothing actually happened. Could give the illusion that something changed when it didn't.

nilsvu committed:
Add dev guide on automatic versioning

nilsvu commented:
@wthrowe @nilsdeppe Pushed fixups and added a dev guide on versioning, that also gives some reasoning. Please take a look!

nilsvu commented:
BTW, this is how the GitHub UI looks like to dispatch a workflow (such screenshots are also in the docs I link in the dev guide)

<img width="479" alt="Bildschirmfoto 2020-11-17 um 18 51 30" src="https://user-images.githubusercontent.com/746230/99427916-45a64800-2906-11eb-9694-48ff1ee79ca6.png">

nilsdeppe commented on tools/CompileReleaseNotes.py:
```diff
@@ -30,7 +32,11 @@ def distance_from_head(tag):
-PullRequest = collections.namedtuple('PullRequest', ['id', 'title'])
+@dataclass
+class PullRequest:
+    id: int
```
I must say, I like that you can now specify types in python (I realize this has been around for a while with Python 3, but I haven't seen it in action until now)

nilsvu replied:
Yep it's nice :) Remember it's only type _annotations_ though, you can still do whatever you want at runtime (for better or for worse)

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,108 @@
+- Python bindings: We try to offer deprecation warnings when changing the Python
+  interfaces to aid developers of external packages that use the Python bindings
+  to keep up.
+
```
Do you think it would be good to state that `We do not support default values in input files, which prevents updating the code and continuing a simulation from subtly changing parameters part way through. Please note that bugs in the actual C++ implementation may alter results, though hopefully this will be a rarity.`? Unclear if in this list, but I think that's something useful to state

nilsvu replied:
added a paragraph in a fixup

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,108 @@
+# Schedule
+
+We typically publish a release at the beginning of each month, and may publish
+additional releases irregularly.
```
Maybe add `If a major bug is discovered and fixed we will likely create a new release.`

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,108 @@
+
+# How to find releases
+
+You find all releases on GitHub:
```
`You CAN find` (add `can`)

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,108 @@
+- [sxs-collaboration/spectre releases on GitHub](https://github.com/sxs-collaboration/spectre/releases)
+
+Each release is tagged with the version name and prefixed `v` in the repository.
+For example, this is how you could obtain the source code for the `2020.11.01`
```
`For example, to obtain the source for for the ...` (rewording to make it less hedging)

nilsvu replied:
how would you continue that sentence so that it connects to the code snippet that comes after?

nilsdeppe commented on docs/DevGuide/Versioning.md:
````diff
@@ -0,0 +1,108 @@
+release:
+
+```sh
+git clone https://github.com/sxs-collaboration/spectre && cd spectre
````
I think `git clone -b v2020.11.01 ...` should work to clone and checkout the specific branch/tag

nilsdeppe commented on docs/DevGuide/Versioning.md:
````diff
@@ -0,0 +1,108 @@
+could obtain the source code for the latest release:
+
+```sh
+git clone https://github.com/sxs-collaboration/spectre && cd spectre
````
`git clone -b release ...` (I know it works for branches :P )

nilsdeppe commented on docs/DevGuide/Versioning.md:
````diff
@@ -0,0 +1,108 @@
+```
+
+The latest release is always available on the `release` branch. This is how you
+could obtain the source code for the latest release:
````
`could` -> `can`?

nilsvu replied:
🗑 Added a can

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,108 @@
+pull-requests. Pull-requests labeled "major new feature" or "bugfix" on GitHub
+will be classified as such in the release notes.
+
+The script `tools/CompileReleaseNotes.py` is responsible for the release notes
```
`is responsible for` -> `generates`?

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,108 @@
+
+The script `tools/CompileReleaseNotes.py` is responsible for the release notes
+and can also be invoked manually with a Python 3 interpreter to retrieve an
+overview on what happened in the repository recently. The script requires you
```
`on what` -> `of what`

nilsdeppe reviewed (commented):
Overall, looks good to me. Just some minor comments. I'm happy for you to squash once @wthrowe has had a chance to look

nilsvu commented:
@nilsdeppe could you check the commit "fixup Add paragraph on input files" before I squash?

nilsdeppe commented on docs/DevGuide/Versioning.md:
```diff
@@ -47,6 +47,16 @@ We may establish guarantees for specific interfaces related to these aspects.
 
+We explicitly provide **no guarantee** that input files remain compatible with
+executables built on different releases. Instead, we try to make input files as
+explicit as possible by not support default values, so simulations don't
```
`support` -> `supporting`

nilsdeppe reviewed (commented):
One small typo

nilsdeppe commented:
Just to make all this even more complicated we will need to support uploading to Zenodo as well... I've figured out how to do this and created a `.zenodo.json` file along with a python script to upload to Zenodo (`tools/ZenodoUpload.py` ) at https://github.com/nilsdeppe/spectre/tree/zenodo.

The really tricky part is that we can't create the DOI after we've released the code because we want the release documentation to have the DOI to cite in it. Basically, in the python script we can tell Zenodo to pre-reserve a DOI, which we can read out and then use to insert into the necessary files. I think if we generalize the python script to be used for all releases we can pass an extra command line argument when we've determined it's time to do a Zenodo release. I think everything in "Edit project version in repo" and "Commit and push" needs to be done inside python. It will also need to wait for GitHub to have created the release so the zip file can be downloaded. Once the zip file is downloaded, it can be uploaded to Zenodo. Having said all that, probably the easiest thing to do for now would be to convert the sed and git commands to be issued from inside a python script (sed could probably just be done within python completely). At that point I think it's a matter of adding another argument to the workflow button that would force Zenodo deployment, and to also have a check to see if the date matches something (e.g. Jan. 1) we also deploy to Zenodo. Does this sound reasonable?

A related question: how secret are the secrets? I'd like to think they're only visible when being run on `develop` (or another branch of the repo itself). Is that correct?

nilsvu commented:
Thanks for looking into the Zenodo upload @nilsdeppe! Would you mind if I take over your commits from https://github.com/nilsdeppe/spectre/tree/zenodo to this branch and integrate them with the versioning?

The Zenodo feature to reserve a DOI before publishing saves us here I think. By far the easiest approach is this:

1. The GitHub action calls Zenodo to reserve a DOI and inserts it into the project along with the version number.
2. The GitHub action creates a release, as is currently does.
3. Zenodo automatically mirrors the release, along with the `.zenodo.json` metadata. We don't have to do anything but enable the Zenodo-GitHub integration.

Edit: The alternative is, as @nilsdeppe describes above, to upload files to Zenodo ourselves, which involves waiting in the script for GitHub to publish the archive. At that point we might as well not automate the Zenodo upload but only reserve a DOI, since we can do a yearly task without automation.

Of course my suggestion means that all releases are mirrored to Zenodo. I think that's the only consistent way to do things anyway, and by far the simplest. Else we end up with a list of releases, only some of which are preserved on Zenodo. The only reason I'm aware of to publish only a subset of releases to Zenodo is to be able to filter citations by year for grant reports. That would be a very useful feature in general, since if this prevents _us_ from fully using Zenodo then it also applies to others, so I think it should be implemented upstream. I created https://github.com/zenodo/zenodo/issues/2076 and volunteered to implement it. This means the feature won't be available immediately, but I think we don't need it right now. @teukolsky I remember you advocated for this feature, so does my suggestion here address your concern?

@nilsdeppe Re your question on secrets: Here's the documentation: https://docs.github.com/actions/reference/encrypted-secrets. Secrets at the repo-level are available to workflows in that repo. For secrets at organization-level you can select the repos they are available to.

nilsvu commented:
@wthrowe could you check the fixups before I squash them?

nilsdeppe commented:
I don't mind if you take it over, but it's not ready to be merged. We still need to double check the info is copied over correctly and have a few remaining comments on the guidelines to address. 

We decided _not_ to use the Zenodo GitHub integration to keep the number of DOIs maintainable.

I don't see why we shouldn't automate the process even if we have to wait for GitHub to generate the release. I assume that generating the release takes less than 10 min, so having the action check if the release is there, sleep for 30 seconds, check again, repeat seems reasonable. What do you have against this? My thought of automating is:

1. Check if the `version` inside Zenodo is of the current year. E.g. we set it to 2019 and then the first GitHub release will be tagged and pushed to Zenodo as 2020.
2. If the version is equal to the current year we do the regular release basically the way you have it now.
3. If the version is lower than the current year we:
  i.  update the Zenodo version in the json file
  ii. reserve a DOI
  iii. update the SpECTRE version and DOI in CMake
  iv. tag and push the new release to GitHub
  v. have the action wait until the release is generated by GitHub (I'm assuming this will take less than an hour even in the worst case scenario)
  vi. download the zip archive GitHub generated and push it to Zenodo
  vii. (optional, but probably useful) post an issue pinging the core-devs list to publish the DOI. 

The main concern was, if I remember correctly, was not that getting the citations for the different DOIs would be too difficult, but that people wouldn't know what DOI to cite if they are constantly changing. Updating on a yearly time scale seemed like a good frequency. I'm not really convinced that revisiting that discussion is worth it at this point since we need a DOI sooner rather than later.

wthrowe commented on tools/CompileReleaseNotes.py:
```diff
@@ -142,51 +146,51 @@ def get_merged_pull_requests(repo: git.Repo, from_rev: str,
+            group_header = {
+                'major new feature': "Major new features",
+                'bugfix': "Bugfixes",
+                'small': "Small changes",
```
If I'm reading correctly, `group` can only be `"major new feature"`, `"bugfix"`, or `None`.

nilsvu replied:
yep, this was left from debugging, removed now

wthrowe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,117 @@
+  processed or converted to newer formats.
+- Python bindings: We try to offer deprecation warnings when changing the Python
+  interfaces to aid developers of external packages that use the Python bindings
+  to keep up.
```
This sentence doesn't seem to make sense.  There are a lot of clauses and they don't seem to connect in a reasonable way.  I'm not quite sure what it is trying to say, but maybe "aid" -> "help" and "to keep" -> "keep"?

nilsvu replied:
reworded

wthrowe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,117 @@
+(and often must) be reviewed before the simulation continues. Please note that
+low-level changes or bugs may always alter results, so be advised to stick with
+a particular compiled executable if exact reproducibility is crucial for your
+project.
```
We don't guarantee exact reproducibility.

nilsvu replied:
also reworded. (or should I remove / change this?)

wthrowe replied:
I think the new version is fine.

wthrowe commented on docs/DevGuide/Versioning.md:
```diff
@@ -0,0 +1,117 @@
+versioning is to avoid implying a notion of compatibility between releases (see
+section below).
+
+# Guarantees and non-guarantees of releases
```
We don't really have a good rule for what goes in the devguide and what goes elsewhere, but I think the devguide is intended for at least somewhat technical and somewhat internal information.  This is about as far as we're ever likely to get from that.  (It is addressed to *users*.)

nilsvu replied:
Agreed. It's not a user tutorial either though.. do you have an idea where to put it? Or should it be split somehow?

wthrowe replied:
I think a split makes sense.  Everything from "Automatic versioning" downwards is really technical, some of it only usable by core devs or something like that.  That seems like a dev guide entry.  The stuff above that is aimed at users.  Maybe link it from the installation page?

wthrowe reviewed (changes requested):
Re identical releases: I guess the question is whether we consider the monthly releases to have some kind of preferred status or whether they are more like "We want to make sure we don't go too long between releases and this is easier than automatically doing one four weeks after the last release."  If the former then a manual release shouldn't prevent a monthly from happening, but if the latter then it is fine if one does so.

nilsvu commented on .github/workflows/Tests.yaml:
```diff
@@ -433,3 +447,94 @@ jobs:
+            exit 1
+          fi
+          if [ $(git rev-parse HEAD) == $(git rev-parse origin/release) ]; then
+            echo "Nothing changed since last release $(git describe release)."
```
@wthrowe moving the discussion on identical releases to a thread so it's easier to follow

> https://github.com/sxs-collaboration/spectre/pull/2598#pullrequestreview-533812299
> Re identical releases: I guess the question is whether we consider the monthly releases to have some kind of preferred status or whether they are more like "We want to make sure we don't go too long between releases and this is easier than automatically doing one four weeks after the last release." If the former then a manual release shouldn't prevent a monthly from happening, but if the latter then it is fine if one does so.

With the version format YYYY.MM.DD (i.e. we never omit the day) I lean towards the latter, i.e. no special status for monthly (or any) releases.

nilsdeppe replied:
Since we don't plan on doing any extra testing for monthly releases I don't think we should consider any release "better" or "special" compared to any other release.

wthrowe replied:
OK, let's go with that and leave the check in.

nilsdeppe commented:
We need to wait to do releases until after #2599 is merged because otherwise people can't build our releases :(

nilsvu commented:
@nilsdeppe good call 😄

nilsvu commented:
@nilsdeppe @wthrowe I added three commits that integrate Zenodo into the workflow. Here's what still needs to be done:

- [x] Read version in CMake from `Metadata.yaml` so we only have to specify it once

If you get the chance please take a look through the additions and the fixups and let me know if I can squash the fixups up to here. Then I can polish everything a bit more.

wthrowe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,550 @@
+    except requests.exceptions.HTTPError as err:
+        raise requests.exceptions.HTTPError((
+            "Are you re-running the script over a repository that already has "
+            "the unpublished new version ID inserted into Metadata.yaml?"
```
Is this suggesting that running over such a repository would cause the error, or that I should be running over such a repository to eliminate the error?

nilsvu replied:
reworded the error message (the former is true)

wthrowe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,550 @@
+            uploaded_file = zenodo.upload_file(bucket_id=bucket_id,
+                                               file=open_tmp_file,
+                                               filename=archive_filename)
+        # with open('/Users/nlf/Projects/spectre/develop/Metadata.yaml',
```
Remove?

wthrowe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,550 @@
+    parent_parser.add_argument(
+        '--zenodo-sandbox',
+        action='store_true',
+        help=("Use the Zenodo sandbox instead of the public version"))
```
Public version of what?

wthrowe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,550 @@
+
+    # @uplink.response_handler(parse_bucket_response)
+    # @uplink.get('deposit/depositions/{id}')
+    # def get_bucket_id(self, deposition_id: uplink.Path(name='id')):
```
Remove?

wthrowe reviewed (changes requested):
I am fine with a squash (including these comments).

nilsdeppe commented on docs/DevGuide/DevGuide.md:
```diff
@@ -57,3 +57,8 @@ Assumes a thorough familiarity and fluency in SpECTRE's usage of TMP.
+### Continuous Integration
+Explanations on our automated tests and deployments can be found here.
+
+- \ref dev_guide_automatic_versioning
```
should there also be a reference to travis here? I know there's one above, but I could see someone looking in this section for it as well

nilsvu replied:
Yes, I'll move it down and update it to refer to GitHub Actions in a separate PR

nilsdeppe replied:
Sounds good, thanks!

nilsdeppe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,580 @@
+    logger.info("CHECK ONLY: {}".format(msg))
+
+
+def parse_new_version_response(response):
```
the name doesn't match the documentation. `new_version_id_from_response`?

nilsdeppe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,580 @@
+
+
+def get_response_text(response):
+    return response.text
```
Sorry, why is this wrapper needed? It seems like calling the wrapper is more code than getting the text

nilsvu replied:
It's a "response handler" function, used for `Github.render_markdown_raw`. Replaced with `operator.attrgetter`.

nilsdeppe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,580 @@
+    publication_date = '{}-{}-{}'.format(*match_version_name.groups()[:3])
+
+    if update_only:
+        # The new version already exists on Zenodo, and we assume the metadata
```
Sorry, I don't really understand the difference between the if and else branch. I think this boils down to me not understanding what `update_only` means. What is being updated?

nilsvu replied:
This is the help string for `update_only`:

> Only update an existing version draft on Zenodo, not creating a new one. Use this flag if the metadata in the repository already reference the new version draft on Zenodo.

nilsdeppe replied:
Sorry, where is the help string located? I tried searching this file and couldn't find it.

nilsvu replied:
I meant this one:

```
    parser_prepare.add_argument(
        '--update-only',
        action='store_true',
        help=(
            "Only update an existing version draft on Zenodo, not creating a "
            "new one. Use this flag if the metadata in the repository already "
            "reference the new version draft on Zenodo."))
```

nilsdeppe replied:
Okay, since I understand that message but neither of the comments related to the if-else could you reword the code comments? I think the thing that confuses me is `exists on Zenodo` sounds like we have hit a bug and should abort.

nilsvu replied:
reworded in a fixup

nilsdeppe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,580 @@
+    ).format(repo.working_dir))
+    parent_parser = argparse.ArgumentParser(add_help=False)
+    parent_parser.add_argument(
+        '--zenodo-token',
```
Question: is this the recommended way of getting tokens into scripts in CI? Mike had suggested retrieving the env variable internally to accidental printing of the API token.

nilsvu replied:
I think this is an OK way of passing secrets. Env vars also work. It's probably best to support both, though maybe not necessary for this script that's only called from the GitHub Actions (it's even in `.github/`). Side note: GitHub redacts secrets that are printed in logs, replacing them with `***`.

nilsdeppe replied:
I guess if we used an env var then it wouldn't need to get redacted and so in the odd case that there is a redaction bug we wouldn't leak? [not a change request, just asking/discussing for knowledge :) ]

nilsdeppe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,580 @@
+        action='store_true',
+        help=("Use the Zenodo sandbox instead of the public version"))
+    parent_parser.add_argument(
+        '--github-token',
```
same question about tokens

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -515,3 +548,142 @@ jobs:
+          python3 .github/scripts/Release.py prepare -vv \
+            --version $RELEASE_VERSION \
+            --zenodo-token ${{ secrets.ZENODO_READWRITE_TOKEN }} \
+            --zenodo-sandbox \
```
I guess this will need to be removed?

nilsvu replied:
Yes, removed now

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -515,3 +548,142 @@ jobs:
+        run: |
+          python3 .github/scripts/Release.py publish -vv \
+            --zenodo-token ${{ secrets.ZENODO_PUBLISH_TOKEN }} \
+            --zenodo-sandbox \
```
probably need to remove the `sandbox` here too once we're ready

nilsdeppe commented on .github/workflows/Tests.yaml:
```diff
@@ -515,3 +548,142 @@ jobs:
+          make doc-check
+          make -j2 ExportCoordinates1D
+          ctest -R InputFiles.ExportCoordinates.Input1D.yaml.execute
+      - name: Deploy to gh-pages
```
Question: should we try to keep documentation for old versions? So that, e.g.
```
spectre-code.org/v2020.11.01
```
would give you the documentation for that? I worry that with monthly releases we will quickly have too much data. The dox on my machine is currently 52MB...

nilsvu replied:
Yea that would certainly be nice.. not sure if it's really necessary though when we advise people to use the latest release. I've also been thinking about printing the release notes to the docs..

nilsdeppe replied:
:+1:
Keeping a "Version History" or "Release Notes" would be nice. I don't think that needs to be (or should be done) as part of this PR. I think it's better to get this merged sooner and then see what releases look like and are doing, then adjust what we don't like :)

nilsvu commented:
@nilsdeppe @wthrowe This PR is now rebased and ready for review. I removed the references to the Zenodo Sandbox and tested a few releases on my fork. I think it works :)

Before merging this we'll need to add the following secrets to the repository:

- `ZENODO_READONLY_TOKEN`: no extra permissions
- `ZENODO_READWRITE_TOKEN`: has "deposit:write" permissions
- `ZENODO_PUBLISH_TOKEN`: has "deposit:write" permissions, and once we are satisfied the auto-publishing works, in a few months time perhaps, we can give it "deposit:actions" permissions and add the `--auto-publish` flag to the "publish" command.

I have these secrets already generated, we'll just need to add them.

wthrowe commented on .github/scripts/Release.py:
```diff
@@ -0,0 +1,549 @@
+
+VERSION_PATTERN = r'(\d{4})\.(\d{2})\.(\d{2})(\.\d+)?'
+PUBLICATION_DATE_PATTERN = r'\d{4}-\d{2}-\d{2}'
+DOI_PATTERN = r'10.\d{4,9}/[-._;()/:a-zA-Z0-9]+'
```
Where did this pattern come from?  It doesn't appear to be the allowed values for a DOI, the recommended HTML encoding of DOIs, or the format we intend to use.  As far as I can tell, the valid DOIs are "10.\<anything\>/\<anything\>".

Also the first `.` needs a `\`.

nilsvu replied:
Pushed a fixup with the narrowed-down pattern. This is the pattern used in `CheckMetadata.py` for Zenodo DOIs.

wthrowe commented on .github/scripts/Release.py:
```diff
@@ -20,7 +20,7 @@
 VERSION_PATTERN = r'(\d{4})\.(\d{2})\.(\d{2})(\.\d+)?'
 PUBLICATION_DATE_PATTERN = r'\d{4}-\d{2}-\d{2}'
-DOI_PATTERN = r'10.\d{4,9}/[-._;()/:a-zA-Z0-9]+'
+DOI_PATTERN = r'10\.\d{4,9}/zenodo\.(\d+)'
```
Why the parens?

nilsvu replied:
Copied from `CheckMetadata.py` 🤷‍♂️ want me to remove them?

nilsvu replied:
removed while squashing

nilsdeppe reviewed (commented):
LGTM, please feel free to squash :)

nilsdeppe reviewed (commented):
Yea, that's a lot clearer. Thanks! Please go ahead and squash :)

nilsvu commented:
@kidder Could you add the Zenodo token that I sent you to the repo? Once that's done this can be merged I think.

kidder commented:
@nilsleiffischer tokens added

nilsvu commented:
@kidder Could you please make sure the tokens are named `ZENODO_READONLY_TOKEN`, `ZENODO_READWRITE_TOKEN` and `ZENODO_PUBLISH_TOKEN`?

kidder commented:
yes, they are named so and have been added as repository secrets

nilsvu commented:
Aha, pull requests don't have access to the repo's secrets (which makes sense). This means we can't do Zenodo checks on PRs (wouldn't be super useful anyway). I updated the workflow.

wthrowe commented:
The new script is erroring in CI.

nilsvu committed:
Don't try to print active branch in release notes script

No active branch is available in detached state. Also, the active branch
is not really relevant for the script. Instead, added a note that the
'--to' argument defaults to 'HEAD'.

nilsvu commented:
Added a small commit at the beginning that removes printing the `repo.active_branch` in the release notes script. The reason this failed is that _only on PRs_ the checkout is in detached state, so the active branch is unavailable. Therefore it didn't show up when pushed on my fork. It's good to have the CI checks 😄

nilsvu commented:
@nilsdeppe @wthrowe @kidder I decided to disable the `schedule` trigger for now to give us some time to test the automatic releases without having to do that on Jan 1st. We can just dispatch the release workflow manually anytime we want to create a release. I suggest Monday Jan 4th.

Could you take another look and see if this is ready to merge?

nilsvu commented:
FYI, GitHub [protected environments](https://docs.github.com/en/free-pro-team@latest/actions/reference/environments) are in public beta right now. Those will allow us to review and approve releases before they get sent to Zenodo.
